packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1216 @@
+  /// bottom of the viewport.
+  AxisDirection get verticalAxisDirection => _verticalAxisDirection;
+  AxisDirection _verticalAxisDirection;
+  set verticalAxisDirection(AxisDirection value) {
+    if (_verticalAxisDirection == value) {
+      return;
+    }
+    _verticalAxisDirection = value;
+    markNeedsLayout();
+  }
+
+  /// Supplies children for layout in the viewport.
+  TwoDimensionalChildDelegate get delegate => _delegate;
+  TwoDimensionalChildDelegate _delegate;
+  set delegate(TwoDimensionalChildDelegate value) {
+    if (_delegate == value) {
+      return;
+    }
+    if (attached) {
+      _delegate.removeListener(_handleDelegateNotification);
+    }
+    final TwoDimensionalChildDelegate oldDelegate = value;
+    _delegate = value;
+    if (attached) {
+      _delegate.addListener(_handleDelegateNotification);
+    }
+    if (_delegate.runtimeType != oldDelegate.runtimeType || _delegate.shouldRebuild(oldDelegate)) {
+      _handleDelegateNotification();
+    }
+  }
+
+  /// The major axis of the two dimensions.
+  ///
+  /// This is can be used by subclasses to determine paint order,
+  /// visitor patterns like row and column major ordering, or hit test
+  /// precedence.
+  ///
+  /// See also:
+  ///
+  ///  * [TwoDimensionalScrollView], which assigns the [PrimaryScrollController]
+  ///    to the [TwoDimensionalScrollView.mainAxis] and shares this value.
+  Axis  get mainAxis => _mainAxis;
+  Axis _mainAxis;
+  set mainAxis(Axis value) {
+    if (_mainAxis == value) {
+      return;
+    }
+    _mainAxis = value;
+    markNeedsLayout();
+  }
+
+  /// {@macro flutter.rendering.RenderViewportBase.cacheExtent}
+  double  get cacheExtent => _cacheExtent ?? RenderAbstractViewport.defaultCacheExtent;
+  double? _cacheExtent;
+  set cacheExtent(double? value) {
+    if (_cacheExtent == value) {
+      return;
+    }
+    _cacheExtent = value;
+    markNeedsLayout();
+  }
+
+  /// {@macro flutter.material.Material.clipBehavior}
+  Clip get clipBehavior => _clipBehavior;
+  Clip _clipBehavior;
+  set clipBehavior(Clip value) {
+    if (_clipBehavior == value) {
+      return;
+    }
+    _clipBehavior = value;
+    markNeedsLayout();
+  }
+
+  final _TwoDimensionalChildManager _childManager;
+  bool _hasVisualOverflow = false;
+  final LayerHandle<ClipRectLayer> _clipRectLayer = LayerHandle<ClipRectLayer>();
+
+  @override
+  bool get isRepaintBoundary => true;
+
+  @override
+  bool get sizedByParent => true;
+
+  /// All of the children in the order that they will be painted.
+  ///
+  /// The [mainAxis] correlates with each [ChildVicinity] to paint in a row or
+  /// column major order.
+  ///
+  /// By default, the [mainAxis] is [Axis.vertical], which would result in a
+  /// row major paint order, visiting children in the horizontal indices before
+  /// advancing to the next vertical index.
+  List<RenderBox> get childrenInPaintOrder => _childrenInPaintOrder;
```

Comment:
Yeah, I don't love it, but wanted to ensure a couple of things

- as a copy of _children (the source of truth), I thought it might be a (little) safer - I figured I would add some assertions that checks childrenInPaintOrder against _children when it is used..
- subclasses could override paint and I wanted to make it clear that children should be painted in this order so we do not violate the expectation when visiting for semantics
- This made overriding hit testing in a subclass easier (but I can change that by exposing a method that checks a hit position against a given child)
- during layoutChildSequence, subclasses can build children in any order they want, but that may not be the order according to childVicinity, so they need to be reified after layout

`getChildFor` is actually an alternative I was working with that is still here.. maybe we could add a `getChildForPaint` method?

Reply:
What I could do is have a getter for firstChild, and instead of setting up a list of children in paint order, I could set up next/previous siblings in parent data...

Reply:
> What I could do is have a getter for firstChild, and instead of setting up a list of children in paint order, I could set up next/previous siblings in parent data...

That's what we do in the canonical ContainerRenderObjectMixin, so that would fit flutter's style pretty well. I like that idea.

Reply:
I will make it so!